pkg/gofr/datasource/pubsub/nats/subscription_manager.go

```diff
@@ -0,0 +1,229 @@
+		return buffer
+	}
+
+	buffer := make(chan *pubsub.Message, sm.bufferSize)
+	sm.topicBuffers[topic] = buffer
+
+	return buffer
+}
+
+func (*SubscriptionManager) createOrUpdateConsumer(
+	ctx context.Context, js jetstream.JetStream, topic string, cfg *Config) (jetstream.Consumer, error) {
+	consumerName := fmt.Sprintf("%s_%s", cfg.Consumer, strings.ReplaceAll(topic, ".", "_"))
+	cons, err := js.CreateOrUpdateConsumer(ctx, cfg.Stream.Stream, jetstream.ConsumerConfig{
+		Durable:       consumerName,
+		AckPolicy:     jetstream.AckExplicitPolicy,
+		FilterSubject: topic,
+		MaxDeliver:    cfg.Stream.MaxDeliver,
+		DeliverPolicy: jetstream.DeliverNewPolicy,
+		AckWait:       30 * time.Second,
+	})
+
+	return cons, err
+}
+
+func (sm *SubscriptionManager) consumeMessages(
+	ctx context.Context,
+	cons jetstream.Consumer,
+	topic string,
+	buffer chan *pubsub.Message,
+	cfg *Config,
+	logger pubsub.Logger) {
+	// TODO: propagate errors to caller
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if err := sm.fetchAndProcessMessages(ctx, cons, topic, buffer, cfg, logger); err != nil {
+				logger.Errorf("Error fetching messages for topic %s: %v", topic, err)
+			}
+		}
+	}
+}
+
+func (sm *SubscriptionManager) fetchAndProcessMessages(
+	ctx context.Context,
+	cons jetstream.Consumer,
+	topic string,
+	buffer chan *pubsub.Message,
+	cfg *Config,
+	logger pubsub.Logger) error {
+	msgs, err := cons.Fetch(1, jetstream.FetchMaxWait(cfg.MaxWait))
+	if err != nil {
+		return sm.handleFetchError(err, topic, logger)
+	}
+
+	return sm.processFetchedMessages(msgs, topic, buffer, logger)
+}
+
+func (sm *SubscriptionManager) handleFetchError(err error, topic string, logger pubsub.Logger) error {
+	if !errors.Is(err, context.DeadlineExceeded) {
+		logger.Errorf("Error fetching messages for topic %s: %v", topic, err)
+	}
+
+	time.Sleep(consumeMessageDelay)
+
+	return nil
+}
+
+func (sm *SubscriptionManager) processFetchedMessages(
+	msgs jetstream.MessageBatch,
+	topic string,
+	buffer chan *pubsub.Message,
+	logger pubsub.Logger) error {
+	for msg := range msgs.Messages() {
+		pubsubMsg := sm.createPubSubMessage(msg, topic)
+
+		if !sm.sendToBuffer(pubsubMsg, buffer, topic, logger) {
+			logger.Logf("Message buffer is full for topic %s. Consider increasing buffer size or processing messages faster.", topic)
+		}
+	}
+
+	return sm.checkBatchError(msgs, topic, logger)
+}
+
+func (sm *SubscriptionManager) createPubSubMessage(msg jetstream.Msg, topic string) *pubsub.Message {
+	pubsubMsg := pubsub.NewMessage(context.Background()) // Pass a context if needed
+	pubsubMsg.Topic = topic
+	pubsubMsg.Value = msg.Data()
+	pubsubMsg.MetaData = msg.Headers()
+	pubsubMsg.Committer = &natsCommitter{msg: msg}
+	return pubsubMsg
+}
+
+func (sm *SubscriptionManager) sendToBuffer(msg *pubsub.Message, buffer chan *pubsub.Message, topic string, logger pubsub.Logger) bool {
+	select {
+	case buffer <- msg:
+		return true
+	default:
+		return false
+	}
```

Comment:
Isn't a simple ?


```suggestion
	_, ok := buffer <- msg:
	return ok
```

Or something like that?

Reply:
IMHO, this syntax will not work as when we do:

```go
ok := buffer <- msg:
return ok
```

We get error: Invalid operation: <- msg (receive from the non-chan type *pubsub.Message)